src/services/gcs/backend.rs

```diff
@@ -0,0 +1,718 @@
+use std::collections::{HashMap, HashSet};
+use std::fmt::{Debug, Formatter};
+use std::io::{Error, ErrorKind, Result};
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use http::StatusCode;
+use isahc::AsyncBody;
+use log::{debug, error, info, warn};
+use metrics::increment_counter;
+use reqsign::services::google::Signer;
+
+use crate::accessor::AccessorCapability;
+use crate::error::{other, BackendError, ObjectError};
+use crate::http_util::{
+    new_http_channel, parse_content_length, parse_error_kind, parse_error_response,
+    parse_error_status_code, parse_etag, parse_last_modified, percent_encode_path, HttpBodyWriter,
+    HttpClient,
+};
+use crate::ops::{
+    BytesRange, OpCreate, OpDelete, OpList, OpPresign, OpRead, OpStat, OpWrite, Operation,
+    PresignedRequest,
+};
+use crate::services::gcs::dir_stream::DirStream;
+use crate::AccessorMetadata;
+use crate::{Accessor, BytesReader, BytesWriter, DirStreamer, ObjectMetadata, ObjectMode, Scheme};
+
+const DEFAULT_GCS_ENDPOINT: &'static str = "https://storage.googleapis.com";
+const DEFAULT_GCS_AUTH: &'static str = "https://www.googleapis.com/auth";
+
+// TODO: Server side encryption support
+
+/// GCS storage backend builder
+#[derive(Clone)]
+pub struct Builder {
+    /// root URI, all operations happens under `root`
+    root: Option<String>,
+    /// bucket name
+    bucket: String,
+    /// endpoint URI of GCS service,
+    /// default is "https://storage.googleapis.com"
+    endpoint: Option<String>,
+
+    /// credential string for GCS service
+    credentials: Option<String>,
+}
+
+impl Default for Builder {
+    fn default() -> Self {
+        Self {
+            root: None,
+            bucket: "".to_string(),
+            endpoint: None,
+            credentials: None,
+        }
+    }
+}
+
+impl Builder {
+    pub fn root(&mut self, root: &str) -> &mut Self {
+        if root.is_empty() {
+            self.root = None;
+        } else {
+            self.root = Some(root.to_string());
+        }
```

Comment:
```suggestion
        if !root.is_empty() {
            self.root = Some(root.to_string());
        }
```

Less is more.

Reply:
cc @ClSlaid

Reply:
This will make passing an empty string do nothing.

I think passing an empty string purging this field, instead of doing nothing, is correct.